File: server/src/Marshaller/ClientPacket.java
package src.marshaller;

/**
 * Class to store a client's request
 * @author Lee Juin
 * @version 1.0
 */
public class ClientPacket {
    /**
     * The request ID of the client
     */
    private int requestID;
    /**
     * The service ID of the client
     */
    private int serviceID;
    /**
     * The length of the request
     */
    private int requestLength;
    /**
     * The path to the file
     */
    private String filePath;
    /**
     * The payload for the service
     */
    private ClientPayload clientPayload;

    /**
     * Constructor for Client Packet
     * @param requestID request ID
     * @param serviceID service ID
     * @param filePath  path to file
     * @param clientPayload payload for the service
     */
    public ClientPacket(int requestID, int serviceID, String filePath, ClientPayload clientPayload) {
        this.requestID = requestID;
        this.serviceID = serviceID;
        this.filePath = filePath;
        this.clientPayload = clientPayload;
    }

    /**
     * Get the request ID
     * @return  The request ID
     */
    public int getRequestID() {
        return this.requestID;
    }

    /**
     * Get the request length
     * @return  The request length
     */
    public int getRequestLength() {
        return this.requestLength;
    }

    /**
     * Get the service ID
     * @return  The service ID
     */
    public int getServiceID() {
        return this.serviceID;
    }

    /**
     * Get the file path
     * @return  The file path
     */
    public String getFilePath() {
        return this.filePath;
    }

    /**
     * Get the client payload
     * @return  The client payload
     */
    public ClientPayload getClientPayload() {
        return this.clientPayload;
    }
}

File: client/src/ClientPacket.cpp
#include "ClientPacket.h"

/**
 * @brief Constructs a ClientPacket object.
 *
 * This constructor initializes a ClientPacket object with the provided
 * parameters.
 *
 * @param requestId The ID of the request.
 * @param serviceId The ID of the service.
 * @param filepath The filepath associated with the packet.
 * @param payload A pointer to the ClientPayload object.
 */
ClientPacket::ClientPacket(int requestId, int serviceId,
                           const std::string& filepath, ClientPayload* payload)
    : requestId(requestId),
      serviceId(serviceId),
      filepath(filepath),
      payload(payload) {}

/**
 * @brief Gets the request ID of the ClientPacket.
 *
 * @return The request ID.
 */
int ClientPacket::getRequestId() const { return requestId; }

/**
 * @brief Gets the service ID of the ClientPacket.
 *
 * @return The service ID.
 */
int ClientPacket::getServiceId() const { return serviceId; }

/**
 * @brief Gets the filepath associated with the ClientPacket.
 *
 * @return The filepath.
 */
const std::string& ClientPacket::getFilepath() const { return filepath; }

/**
 * @brief Gets the payload of the ClientPacket.
 *
 * @return A pointer to the ClientPayload object.
 */
ClientPayload* ClientPacket::getPayload() const { return payload; }

// Setters
/**
 * @brief Sets the request ID of the ClientPacket.
 *
 * @param requestId The request ID to set.
 */
void ClientPacket::setRequestId(int requestId) { this->requestId = requestId; }

/**
 * @brief Sets the service ID of the ClientPacket.
 *
 * @param serviceId The service ID to set.
 */
void ClientPacket::setServiceId(int serviceId) { this->serviceId = serviceId; }

/**
 * @brief Sets the filepath associated with the ClientPacket.
 *
 * @param filepath The filepath to set.
 */
void ClientPacket::setFilepath(const std::string& filepath) {
    this->filepath = filepath;
}

/**
 * @brief Sets the payload of the ClientPacket.
 *
 * @param payload A pointer to the ClientPayload object to set.
 */
void ClientPacket::setPayload(ClientPayload* payload) {
    this->payload = payload;
}